torch/csrc/jit/passes/lower_tuples.cpp

```diff
@@ -8,17 +8,55 @@ namespace torch { namespace jit {
 namespace {
 
 // operators where we expect to find tuples as inputs/outputs
-// this is to assert we are only  doing modifications when we know
+// this is to assert we are only doing modifications when we know
 // we can flatten tuples
 std::unordered_set<Symbol> white_list = {
   prim::If,
   prim::Loop,
   prim::TupleUnpack,
   prim::TupleConstruct,
+  prim::TupleIndex,
+  prim::TupleSlice,
   prim::Param,
   prim::Return,
 };
 
+// returns true if a tuple operator was removed
+bool removeTupleNodes(Node *n, bool must_remove_tuples) {
+  if (n->kind() != prim::TupleUnpack && n->kind() != prim::TupleIndex
+      && n->kind() != prim::TupleSlice) {
+    return false;
+  }
+  auto construct = n->input()->node();
+  if (construct->kind() != prim::TupleConstruct) {
+    if (must_remove_tuples) {
+      AT_ERROR(n->kind().toQualString(), " not matched to tuple construct");
+    }
+    return false;
+  }
+  if (n->kind() == prim::TupleUnpack) {
+    for(size_t i = 0; i < n->outputs().size(); ++i) {
+      n->outputs()[i]->replaceAllUsesWith(construct->inputs()[i]);
+    }
+  } else if (n->kind() == prim::TupleIndex) {
+    auto idx = n->i(attr::index);
+    n->output()->replaceAllUsesWith(construct->inputs()[idx]);
```

Comment:
.at

Reply:
(otherwise you are setting up some inscrutable crashes when someone gives you an invalid graph)